rpc-re-assembler/assembleRPCs.js

```diff
@@ -0,0 +1,139 @@
+    const orderedClientData = _.orderBy(clientData, ['sequenceNumber'], ['asc']);
+    const orderedTargetData = _.orderBy(targetData, ['sequenceNumber'], ['asc']);
+    return {
+        orderedClientData,
+        orderedTargetData,
+    };
+}
+
+function splitClientTargetData(jsonFiles) {
+    const clientFiles = [];
+    const targetFiles = [];
+    jsonFiles.forEach((file) => {
+        if (file.type === 'client') {
+            clientFiles.push(file);
+        } else {
+            targetFiles.push(file);
+        }
+    });
+    return {
+        clientFiles,
+        targetFiles,
+    };
+}
+
+function getRawData(data) {
+    const rawData = data.map(obj => obj.data).join('');
+    const splitRawData = rawData.split(DELIMITER);
+    // preserve the '\u0004' delimiters for the final output
+    const delimited = splitRawData.map(splitData => `${splitData}${DELIMITER}`);
+    return delimited;
+}
+
+function parseRawData(clientData, targetData) {
+    const parsedClientData = clientData.map(data => ({ raw: data, parsed: rpcParser.parseRawRPC(data) }));
+    const parsedTargetData = targetData.map(data => ({ raw: data, parsed: rpcParser.parseRawResults(data) }));
+    return {
+        parsedClientData,
+        parsedTargetData,
+    };
+}
+
+// 1. split client/target files
+// 2. flatten files into two arrays of kinesis.data objects
+// 3. sort the arrays separately by sequenceNumber
+// 4. split raw data by '\u0004' delimiter (but preserve delimiters)
+// 5. parse raw RPC data with rpcParser util
+function parseRPCData(jsonFiles) {
+    const splitJsonFiles = splitClientTargetData(jsonFiles);
+    const { clientFiles, targetFiles } = splitJsonFiles;
+    let clientData = [];
+    let targetData = [];
+    clientFiles.forEach((file) => {
+        clientData = _.concat(clientData, file.kinesis.data);
+    });
+    targetFiles.forEach((file) => {
+        targetData = _.concat(targetData, file.kinesis.data);
+    });
+    const orderedFiles = orderData(clientData, targetData);
+    const { orderedClientData, orderedTargetData } = orderedFiles;
+    const rawClientData = getRawData(orderedClientData);
+    const rawTargetData = getRawData(orderedTargetData);
+    const { parsedClientData, parsedTargetData, } = parseRawData(rawClientData, rawTargetData);
+    const combinedData = combineClientTargetData(parsedClientData, parsedTargetData);
+    writeOutputToFile(combinedData);
+}
+
+// Asyncronously read directory and files, then parse the data
+async function readFiles(filenames) {
+    const promises = filenames.map(async (fileName) => {
+        const type = fileName.split('_')[1];
+        const contents = await fs.readFile(`${SAMPLE_DIR}/${fileName}`, 'utf-8');
+        const jsonFile = JSON.parse(contents);
+        jsonFile.type = type;
+        return jsonFile;
+    });
+    await Promise.all(promises).then((jsonFiles) => {
+        parseRPCData(jsonFiles);
+    }).catch((err) => {
+        console.log(err);
+    });
```

Comment:
Consider taking fuller advantage of the async/await syntax for a better, cleaner way to handle the Promise execution flow:
```suggestion
try {
    const jsonFiles = await Promise.all(promises);
    parseRPCData(jsonFiles);
} catch (err) {
    console.log(err);
}
```

Reply:
Pushed up some code that we can discuss tomorrow when you have time. Sometimes it works, sometimes it doesn't...
Still working on creating a robust set of tests.

Reply:
Okay...yeah, you'll have to explain what "sometimes it works, sometimes it doesn't" means... 😆